File: whiteboard/app/src/main/java/com/ditto/whiteboard/transport/InMemoryWhiteboardTransport.kt
package com.ditto.whiteboard.transport

import com.ditto.whiteboard.domain.BoardOperation
import com.ditto.whiteboard.domain.LivePreview
import com.ditto.whiteboard.domain.UserProfile
import java.util.UUID
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow

class InMemoryWhiteboardTransport(
  override val localPeerKey: String = "local-${UUID.randomUUID().toString().take(8)}",
  private val reason: String,
) : WhiteboardTransport {
  private val mutableEvents = MutableSharedFlow<TransportEvent>(extraBufferCapacity = 64)
  override val events = mutableEvents.asSharedFlow()
  private val mutableDiagnostics = MutableStateFlow(
    TransportDiagnostics(
      localPeerKey = localPeerKey,
      mode = TransportMode.LocalPreview,
      connectivityMessage = reason,
    ),
  )
  override val diagnostics = mutableDiagnostics.asStateFlow()
  private var foreground = true
  private val knownLog = BoundedOperationLog()

  override suspend fun start(profile: UserProfile) {
    mutableDiagnostics.value = mutableDiagnostics.value.copy(running = foreground)
  }

  override suspend fun sendReliable(operation: BoardOperation): Boolean =
    knownLog.accept(operation) !is KnownOperationResult.Rejected

  override fun sendLive(preview: LivePreview) = Unit

  override fun setForeground(isForeground: Boolean) {
    foreground = isForeground
    mutableDiagnostics.value = mutableDiagnostics.value.copy(running = isForeground)
  }

  override fun close() {
    mutableDiagnostics.value = mutableDiagnostics.value.copy(running = false)
  }
}

File: whiteboard-apple/Packages/WhiteboardCore/Sources/WhiteboardCore/Transport/InMemoryWhiteboardTransport.swift
import Foundation

/// Local-only fallback transport used when credentials or the Ditto SDK are unavailable. Accepts
/// operations into a `BoundedOperationLog` so the same terminal limits apply as in nearby mesh;
/// previews go nowhere and the mode is always `localPreview`.
public final class InMemoryWhiteboardTransport: WhiteboardTransport, @unchecked Sendable {
  public let localPeerKey: String

  private let lock = NSLock()
  private var foreground = true
  private var knownLog = BoundedOperationLog()
  private var diagnosticsValue: TransportDiagnostics
  private let eventsBroadcast = AsyncBroadcast<TransportEvent>()
  private let diagnosticsBroadcast: AsyncBroadcast<TransportDiagnostics>

  public init(
    localPeerKey: String = "local-\(UUID().uuidString.prefix(8).lowercased())",
    reason: String
  ) {
    self.localPeerKey = localPeerKey
    let initial = TransportDiagnostics(
      localPeerKey: localPeerKey,
      mode: .localPreview,
      connectivityMessage: reason
    )
    diagnosticsValue = initial
    diagnosticsBroadcast = AsyncBroadcast(replayLatest: true, latest: initial)
  }

  public var events: AsyncStream<TransportEvent> { eventsBroadcast.stream }

  public var diagnostics: AsyncStream<TransportDiagnostics> { diagnosticsBroadcast.stream }

  public var currentDiagnostics: TransportDiagnostics {
    lock.withLock { diagnosticsValue }
  }

  public func start(profile: UserProfile) async {
    let running = lock.withLock { foreground }
    updateDiagnostics { $0.running = running }
  }

  public func sendReliable(_ operation: BoardOperation) async -> Bool {
    let result = lock.withLock { knownLog.accept(operation) }
    if case .rejected = result { return false }
    return true
  }

  public func sendLive(_ preview: LivePreview) {}

  public func setForeground(_ isForeground: Bool) {
    lock.withLock { foreground = isForeground }
    updateDiagnostics { $0.running = isForeground }
  }

  /// Local-only preview: the graph is always just this device. Fires once so the presence
  /// viewer renders the "Me" pill, then stays silent.
  public func observePresenceGraph(
    _ handler: @escaping @Sendable (PresenceGraphSnapshot) -> Void
  ) -> PresenceGraphObservationToken {
    let snapshot = PresenceGraphSnapshot(
      localPeer: PresencePeerSnapshot(peerKey: localPeerKey, deviceName: ""),
      remotePeers: []
    )
    Task { handler(snapshot) }
    return PresenceGraphObservationToken()
  }

  public func syncStatusByPeerKey() async -> [String: PeerSyncStatus] { [:] }

  public func close() {
    updateDiagnostics { $0.running = false }
    eventsBroadcast.finish()
  }

  private func updateDiagnostics(_ mutate: (inout TransportDiagnostics) -> Void) {
    let next: TransportDiagnostics? = lock.withLock {
      mutate(&diagnosticsValue)
      return diagnosticsValue
    }
    // MutableStateFlow conflates equal values; keep the same emission discipline.
    if let next, next != diagnosticsBroadcast.latestSnapshot {
      diagnosticsBroadcast.yield(next)
    }
  }
}
